### PR TITLE
Return to Lobby using Hypersleep.

### PR DESCRIPTION
**Pull request by ActualRory**
Vote

Implements the above.

You can now return to the lobby using Hypersleep Cryopods.

### CODE REVIEW

**Comment by Ghommie** [comment]
Sounds ok, but shouldn't a "limit" to how many times a given player can return to the lobby be coded there?
Just to limit the aforementioned cons of this function.

**Comment by ActualRory** [comment]
> Sounds ok, but shouldn't a "limit" to how many times a given player can return to the lobby be coded there?
> Just to limit the aforementioned cons of this function. Probably could have other requirements in future though.

This is a good idea. I will add this.

File: code/game/machinery/cryopod.dm [hunk 1]
```diff
@@ -187,229 +187,231 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+	announce.autosay("[occupant.real_name] has entered long-term hypersleep storage. Belongings moved to hypersleep inventory.", "Hypersleep Storage System")
+	visible_message("<span class='notice'>[src] hums and hisses as it moves [occupant.real_name] into hypersleep storage.</span>")
 
-			cdel(occupant)
-			occupant = null
-			stop_processing()
+	//Delete the mob.
+	var/mob/new_player/NP = new()
+	NP.ckey = occupant.ckey
+	if(NP.client) NP.client.change_view(world.view)
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]:
```suggestion
	NP.client?.change_view(world.view)
```

Comment (Rohesie) [reply code/game/machinery/cryopod.dm]: Remember not to use single-line control statements in the future, as well.

File: code/game/machinery/cryopod.dm [hunk 2]
```diff
@@ -472,6 +474,23 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
 
+/obj/machinery/cryopod/verb/return_to_lobby()
+	set name = "Return to Lobby"
+	set category = "Object"
+	set src in oview(1)
+	if(occupant != usr)
+		to_chat(usr, "<span class='warning'>You must be inside the [src] to do this.</span>")
+		return
+
+	if(alert(usr, "Do you want to send yourself back to the lobby? This will be logged.","Message","Yes","No") == "Yes")
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]: Do a `== "No"` then `return` to diminish indentation. We call that an early return.

File: code/game/machinery/cryopod.dm [hunk 3]
```diff
@@ -472,6 +474,23 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+		log_admin("[usr.mind.assigned_role] [key_name(usr)] returned to the lobby using a cryopod.")
+		message_admins("\blue [usr.mind.assigned_role] [key_name(usr)] returned to the lobby using a cryopod.")
+		var/mob/new_player/NP = new()
+		NP.ckey = occupant.ckey
+		if(NP.client) NP.client.change_view(world.view)
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]:
```suggestion
		NP.client?.change_view(world.view)
```

File: code/game/machinery/cryopod.dm (hunk 2 above)
Comment (Rohesie) [reply code/game/machinery/cryopod.dm]: After the alert is called the user may no longer be inside the cryo machine. That's something to consider.

**Review by Rohesie — CHANGES_REQUESTED** [review]

File: code/game/machinery/cryopod.dm (hunk 1 above)
Comment (ActualRory) [reply code/game/machinery/cryopod.dm]: Thanks for the tip!

File: code/game/machinery/cryopod.dm (hunk 2 above)
Comment (ActualRory) [reply code/game/machinery/cryopod.dm]: Turns out the return was redundant and not needed at all? Go figure.

**Comment by ActualRory** [comment]
Added respawn tracking and logging.
Hard limit is now three returns.
Soft limit is when admin raises an eyebrow at the notice.

**Comment by Rohesie** [comment]
Make the respawn a client var, and place only one respawn by default, please.

**Comment by ActualRory** [comment]
Will do!

File: code/game/machinery/cryopod.dm [hunk 4]
```diff
@@ -472,6 +474,28 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
 			//Despawning occurs when process() is called with an occupant without a client.
 			add_fingerprint(M)
 
+/obj/machinery/cryopod/verb/return_to_lobby()
+	set name = "Return to Lobby"
+	set category = "Object"
+	set src in oview(1)
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]:
```suggestion
	set src in view(0)
```

Comment (Rohesie) [reply code/game/machinery/cryopod.dm]: `oview(1)` does a 3x3 scan of the object excluding the center.
`view(0)` only includes the center, which is what we want, for this to appear when the user is in the same tile.

File: code/game/machinery/cryopod.dm [hunk 5]
```diff
@@ -472,6 +474,28 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+	if(occupant != usr)
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]: Typecast `usr` here, please. Do we want to send back anything besides humans?
You are running some checks there like the mind one that is not necessarily defined for every possible `usr`.

**Review by Rohesie — CHANGES_REQUESTED** [review]

**Pull request closed without merging**